Source/DafnyCore/Verifier/Translator.ExpressionWellformed.cs

```diff
@@ -895,6 +897,37 @@ private class WFOptions {
                 builder.Add(Assert(GetToken(expr), Bpl.Expr.Le(Bpl.Expr.Literal(0), etran.TrExpr(e.E1)), positiveDesc, options.AssertKv));
                 builder.Add(Assert(GetToken(expr), Bpl.Expr.Le(etran.TrExpr(e.E1), Bpl.Expr.Literal(w)), upperDesc, options.AssertKv));
               }
+            }
+            break;
+          case BinaryExpr.ResolvedOpcode.EqCommon:
+          case BinaryExpr.ResolvedOpcode.NeqCommon:
+            CheckWellformed(e.E1, options, locals, builder, etran);
+            if (e.InCompiledContext) {
+              if (Resolver.CanCompareWith(e.E0) || Resolver.CanCompareWith(e.E1)) {
```

Comment:
As far as _ghost constructors_ are concerned, your hunch is correct. For such types, I believe `CanCompareWith` always returns the same thing for `e.E0` and `e.E1`, so an `&&` would seem sensical and, in fact, it would be fine to check just one of these. However, the use of `CanCompareWith` is to allow comparisons like `x == None`, where `None` is a constructor and the type of `x` is a type that, in general, does not support equality. Such comparisons can be allowed, provided there are no issues of uninitialized values. For the comparison `x == None`, `CanCompareWith(e.E0)` returns `false` whereas `CanCompareWith(e.E1)` returns `true`.

So, we do need both comparisons, and they should be connected with a `||`. I added a test case to make sure this works as intended.